flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingRetrievalBase.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.leaderelection;
+
+import org.apache.flink.runtime.leaderretrieval.LeaderRetrievalListener;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.UUID;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Base class which provides some convenience functions for testing purposes of {@link LeaderRetrievalListener} and
+ * {@link org.apache.flink.runtime.leaderretrieval.LeaderRetrievalEventHandler}.
+ */
+public class TestingRetrievalBase {
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+	private final Object lock = new Object();
+
+	private String address;
+	private String oldAddress;
+	private UUID leaderSessionID;
+	private Exception exception;
+
+
+	public String getAddress() {
+		return address;
+	}
+
+	public UUID getLeaderSessionID() {
+		return leaderSessionID;
+	}
+
+	public String waitForNewLeader(long timeout) throws Exception {
+		long start = System.currentTimeMillis();
+		long curTimeout;
+
+		synchronized (lock) {
+			while (
+				exception == null &&
+					(address == null || address.equals(oldAddress)) &&
+					(curTimeout = timeout - System.currentTimeMillis() + start) > 0) {
+				try {
+					lock.wait(curTimeout);
+				} catch (InterruptedException e) {
+					// we got interrupted so check again for the condition
+				}
+			}
+		}
```

Comment:
I am wondering whether we can't solve this problem a bit more elegantly using a blocking array queue.

Reply:
Same here. This can be a follow up task.